download_generic.go

```diff
@@ -0,0 +1,82 @@
+package main
+
+import (
+	"crypto/md5"
+	"encoding/hex"
+	"fmt"
+	"github.com/hashicorp/go-getter"
+	"github.com/sirupsen/logrus"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+// Calculates the md5sum of a local file
+func md5sum(path string) (string, error) {
+	file, err := os.Open(path)
+	if err != nil {
+		return "", err
+	}
+	defer file.Close()
+
+	hash := md5.New()
+	_, err = io.Copy(hash, file)
+	if err != nil {
+		return "", err
+	}
+
+	byteHash := hash.Sum(nil)[:16]
+	return hex.EncodeToString(byteHash), nil
+}
+
+func idempotentGenericFileDownload(resource, path string) error {
```

Comment:
is generic correct? its only used for s3 at the moment since http still goes through a separate route and apparently also does all the checksum stuff still in `idempotentFileDownload`? Would have expected that to be abstracted out.

Reply:
Although it is only used for S3, it can really be used for anything. I wonder if we should rename the `--s3Obj` arg to something else? WDYT?

Also, I had intended to abstract this out but it would change the behavior (this verifies checksum of the downloaded file, doesn't have the basic auth logic for pulling the checksum, and generally uses a whole different downloading strategy).

Reply:
i think focused change until there is reason to be generic makes sense.

Reply:
Okay I've made the idempotent logic generic now. The only main difference with the existing http-download path is that we check the checksum of the downloaded file to make sure it is correct.

Reply:
Now both the go-getter download path and the custom http downloader have to implement:
```
type downloader interface {
	Download(remotePath, outputPath string) error
	DownloadAndValidateChecksum(remotePath, outputPath, checksum string) error
	RemoteChecksum(remotePath string) (string, error)
}
```